### ISSUE TITLE
@stream directive support

### ISSUE THREAD

**Issue by D1plo1d** [async-graphql/async-graphql#403]
I couldn't find any references to @stream in the docs or issues however I see @defer is supported. Does async-graphql support the @stream directive or could @stream support be added?

Reference: https://github.com/graphql/graphql-wg/issues/329

**Comment by sunli829** [comment]
In the 1.x version, @defer and @stream are supported, but these codes have been removed for some reason. I will add them again after these directives have become standards.

**Comment by D1plo1d** [comment]
I've managed to work around where I was going to use `@stream` so it's no longer a pressing concern for me.

Do you think a feature flag could be appropriate for these? So that developers can choose to opt in to not yet standardized behavior?

I've been helping to explore how `@live` might work in JavaScript and I was hoping to bring that into async-graphql as well.. I don't expect a standard to emerge on `@live` without implementations but if the implementations existed in async-graphql and js then I wonder if a standard can be promoted based on a couple compatible implementations.

**Comment by sunli829** [comment]
> I've managed to work around where I was going to use `@stream` so it's no longer a pressing concern for me.
> 
> Do you think a feature flag could be appropriate for these? So that developers can choose to opt in to not yet standardized behavior?
> 
> I've been helping to explore how `@live` might work in JavaScript and I was hoping to bring that into async-graphql as well.. I don't expect a standard to emerge on `@live` without implementations but if the implementations existed in async-graphql and js then I wonder if a standard can be promoted based on a couple compatible implementations.

In fact, `@live` is very easy to implement for other languages, but not easy for Rust. I hope this feature can be very convenient to use, but I haven't found a suitable way yet.

**Comment by D1plo1d** [comment]
Yes me too. I was thinking of doing a minimal proof of concept using subscriptions and the look-ahead next week - I'll let you know if I find any interesting solutions.

I'm hoping a port of [graphql-live-subscriptions](https://github.com/D1plo1d/graphql-live-subscriptions#client-subscription) will give me [80% of what I want](https://en.wikipedia.org/wiki/Pareto_principle) from `@live`

**Comment by D1plo1d** [comment]
@sunli829 I started working on the `ctx.field().selection_set()` based live subscription and immediately ran into `SelectionField`'s lack of args access. Is this what you meant about not being easy in Rust? Are there lifetime issues with exposing field args through SelectionFeild?

**Comment by sunli829** [comment]
> @sunli829 I started working on the `ctx.field().selection_set()` based live subscription and immediately ran into `SelectionField`'s lack of args access. Is this what you meant about not being easy in Rust? Are there lifetime issues with exposing field args through SelectionFeild?

The main problem is that the receiver of the `OutputType::resolve` function is `&self` not `self`. If I want to save this object and resolve it later, I need to clone it, but if all `OutputType` are required to implement `Clone`, this change is too big.

Maybe I haven't figured out the solution to this problem.

**Comment by D1plo1d** [comment]
In [n1ru4l/graphql-live-query](https://github.com/n1ru4l/graphql-live-query) the `@live` directive is restricted to only operations. So you can have `query @live { things }` but not `query { things @live }`. If we restrict `@live` to the operation in do we get around the need to clone `OutputType`s? Maybe we would need to add clone to only the Query OutputType which is I'm guessing an empty struct for most async-graphql users?

**Comment by D1plo1d** [comment]
The issue that prompted the restriction of `@live` to operators also applies here: `query @live { a: { b } }` is easy to reason about - if a changes then the query will be updated. However `query { a: { b @live } }` is less clear: how do we handle if b is live and not nullable but a is deleted - it is not immediately obvious to me that there is a good solution there.

**Comment by sunli829** [comment]
> In [n1ru4l/graphql-live-query](https://github.com/n1ru4l/graphql-live-query) the `@live` directive is restricted to only operations. So you can have `query @live { things }` but not `query { things @live }`. If we restrict `@live` to the operation in async-graphql do we get around the need to clone `OutputType`s? Maybe we would need to add clone to only the Query OutputType which is I'm guessing an empty struct for most async-graphql users?

If the `@live` directive is restricted to only operations, then it seems easier to implement.

**Comment by bbigras** [comment]
Still relevant.

**Comment by Tehnix** [comment]
It seems `@defer` at least is close to being standardized, with [Apollo Client supporting it from version 3.7](https://www.apollographql.com/docs/react/data/defer/) as well as in their [Apollo Router](https://www.apollographql.com/docs/router/executing-operations/defer-support/) which is written in Rust and might give some inspiration for an implementation.

Would it make sense to split out tracking of `@defer` support from `@stream` since they seem to be adopted at a different pace?

(side-note: I'd almost consider Apollo supporting it as meaning an official standard given its widespread use, but that's very much just IMHO)

**Comment by Zizico2** [comment]
I'd like to try implementing this. Can anyone provide some guidance, or should I just dive in, and figure it out myself?

**Comment by gusinacio** [comment]
Are there any plans for implementation on this?

**Comment by Sytten** [comment]
Reference to the current spec: https://github.com/graphql/graphql-spec/pull/1110
